### PR TITLE
cli: merge op heads and snapshot working copy by "op log" by default

### PR DESCRIPTION
**Pull request by yuja**
# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

File: cli/src/cli_util.rs
Comment (ilyagr) [inline cli/src/cli_util.rs:2436]: (Optional) There is a bit of mixed tenses here: "loads ... they will be merged". One way to phrase this:

I'd reword this a bit:

> By default, Jujutsu loads the repo at the most recent operation, or at the merge of the latest concurrent operations if there is more than one.

I'd also consider adding a paragraph break.

File: cli/src/cli_util.rs
Comment (ilyagr) [inline cli/src/cli_util.rs:2441]: I find it a bit confusing that `--at-op=@` does not seem to imply `--ignore-working-copy`. Is this important, or can we make it more consistent by making it imply `--ignore-working-copy`?

If so, I'd delete this sentence and change one of the paragraphs below to:


> Using `--at-op` has two effects in addition to loading an earlier operation. First, the working copy is ignored, as if `--ignore-working-copy` was specified. Second, `jj` will *not* merge concurrent operations, if any, like it does by default.

File: cli/src/commands/operation/log.rs
Comment (ilyagr) [inline cli/src/commands/operation/log.rs:19]: Perhaps add some docs:

> Show the operation log
>
> Like other `jj` operations, by default `jj op log` may snapshot the working copy or merge most recent concurrent operations. If you'd like to make sure the operation is read-only, use `jj op log --at-op @ --ignore-working-copy`.

Also, if there is concurrency, it seems there is no way to see both concurrent operations in on `jj op log` without merging them, is there? It would be nice if `jj op log --at-op @` showed both (I guess it'd happen at the auto-merge of the concurrent operations without recording the merge), but I'm guessing now it'd be an error message (or did you figure out a better solution?).

If that's correct, I suppose we could add something like

> If there is no unique latest operation, `jj op log --at-op @` will produce an informative error message that gives ids for the most recent concurrent operations.

File: cli/src/cli_util.rs
Comment (ilyagr) [reply cli/src/cli_util.rs:2441]: One  confusing implication (even if possibly rarely in play) would be that if `abc` happens to refer to `@`, then `--at-op=abc` will not imply `--ignore-working-copy`, but once another operation happens, the same option will imply `--ignore-working-copy`.

File: cli/src/cli_util.rs
Comment (yuja) [reply cli/src/cli_util.rs]: Thanks. Updated as
```
    /// Operation to load the repo at. By default, Jujutsu loads the repo at the
    /// most recent operation, or at the merge of the concurrent operations if
    /// any.
    ///
    /// You can use `--at-op=<operation ID>` to see what the repo looked like at
```

File: cli/src/cli_util.rs
Comment (yuja) [reply cli/src/cli_util.rs]: I don't think `--at-op=@` should disable snapshot. And I personally think it's a bug that `--at-op=<head_id>` implies `--ignore-working-copy`.

The idea is that `--at-op=<no_head>` implies `--ignore-working-copy` because the working copy doesn't represent the specified operation.

File: cli/src/commands/operation/log.rs
Comment (yuja) [reply cli/src/commands/operation/log.rs:19]: I think the goal is to remove special case and docs about the special case (such as `op log` doesn't snapshot but `op diff` does.)

Maybe we can add more detailed hint if `op log --at-op=@` failed to load the concurrent heads, but I assume the situation would be pretty rare.

File: cli/src/cli_util.rs
Comment (ilyagr) [reply cli/src/cli_util.rs]: >  I personally think it's a bug that --at-op=<head_id> implies --ignore-working-copy.

I can understand this from a purity standpoint. I find the idea of "fixing" this bug quite confusing though; I can't tell how bizarre the resulting behavior of `jj log --at-op @-` would be. Presumably, the result would be:

```
@ merge
| \
|  o new_snapshot
|  |
o  | former_@
| /
o  former_@-
```

whereas the user would likely expect (if they expect the operation to snapshot at all)

```
@ new_snapshot
o former_@
o former_@-
```

I do not understand merging well enough to understand whether the two resulting states `@` are the same or not. If not, I imagine the difference would be quite confusing.

------------------

Assuming you see a solution to the above problem, I'd be OK with "fixing the bug". If we don't fix the bug, I'd prefer consistently ignoring the working copy always, except perhaps if there is some use-case where snapshotting it is helpful, for example because of the confusing situation that you can't tell whether `jj log --at-op abc` will ignore working copy without knowing what `abc` is.

All that said, I'll ultimately leave this up to you and Martin, you've been discussing this for a while and my understanding of these issues feels a bit superficial at the moment.

File: cli/src/commands/operation/log.rs
Comment (ilyagr) [reply cli/src/commands/operation/log.rs:19]: > Maybe we can add more detailed hint if op log --at-op=@ failed to load the concurrent heads

I think that would be a fine alternative. If `--at-op=@` does not imply `--ignore-working-copy`, that might need a detailed hint too. Ultimately, I'll leave this up to you as well, but here's my motivation.

My main concern is that we might want to pay special attention to commands used to debug or recover from broken states. Knowing which commands are read-only can be critical in this case.

More philosophically, I think that beginner or intermediate `jj` users (especially early adopters) would often get into broken states. We should be motivated to help them, not just out of benevolence, but also because we disproportionately benefit from bug reports such users provide. So, it's in our interest to give them all the guidance we can on how to get meaningful debug info and how to recover. But I may be biased, since I tend to put myself into weird situations often.

File: cli/src/cli_util.rs
Comment (ilyagr) [reply cli/src/cli_util.rs]: Or would `jj log --at-op @-` just fail (if it `--at-op` does not imply `--ignore-working-copy`) because after the snapshot and the merge, `@-` would no longer refer to `former_@-` as the user expects, but refers to two operations?

Comment (yuja) [reply cli/src/cli_util.rs]: Maybe I don't follow your point, but `jj --at-op @-` fails if there are more than one  op heads. That's unchanged. This patch makes it even stricter, `jj --at-op @` now fails.

Oh, maybe you mean `jj --at-op @-` could do snapshot (if `@-` can be resolved to a single operation)? I don't think enabling snapshot makes sense in that situation (though it's technically possible if the working copy tree is identical to the `@-` one.)

What I meant is `jj --at-op @` and `jj --at-op <head-id>` (where `<head-id>` points to the current operation head) don't have to imply `--ignore-working-copy`.

Comment (ilyagr) [reply cli/src/cli_util.rs]: I guess I misunderstood you. My understanding was that you suggested that `--at-op anything` shouldn't imply `--ignore-working-copy`, but I see that was wrong.

I also thought that with your implementation, neither `--at-op @` nor `--at-op <head_id>` implied `--ignore-working-copy`. It sounds like what you were actually saying is that while `--at-op @` *does not* imply `--ignore-working-copy`, `--at-op <head_id>` *does*, and *that*'s what you thought was a bug. 

To me, it would seem that there is no good choice for `--at-op <head_id>` if `--at-op <another_id>` and `--at-op @` have different behavior. If we **had** to choose I'd agree with you that making it behave the same as `--at-op @` is probably better, but it doesn't seem good.

This is why I thought it'd be simpler to make `--at-op <anything>` imply `--ignore-working-copy`. The only downside I see is that it's hard to have the behavior of "do snapshot but don't merge concurrent heads", but I don't know when that would be desired. If there is a use-case, we could perhaps make `jj debug snapshot --at-op <anything>` not `--ignore-working-copy`, which would cover that and other weird cases (for testing snapshotting maybe? I'm unsure).

But again, I could easily be missing something important still.

Comment (yuja) [reply cli/src/cli_util.rs]: > while `--at-op @` _does not_ imply `--ignore-working-copy`, `--at-op <head_id>` _does_, and _that_'s what you thought was a bug.

Correct. It wouldn't practically matter, but seemed a bit odd.

> This is why I thought it'd be simpler to make `--at-op <anything>` imply `--ignore-working-copy`. The only downside I see is that it's hard to have the behavior of "do snapshot but don't merge concurrent heads",

I don't think this operation mode (= don't merge ops, do snapshot) is useful, but I don't think `--at-op=@` should disable snapshot either. These two flags are conceptually different. That said, I'm not strongly against your idea because the implementation will get simpler.

Updated the todo comment.

**Review by martinvonz — APPROVED** [review]
This all looks good to me, FWIW.

File: cli/src/commands/operation/log.rs
Comment (yuja) [reply cli/src/commands/operation/log.rs:19]: Update the help text. Can you take another look?
```
/// Like other commands, `jj op log` snapshots the current working-copy changes
/// and merges concurrent operations. Use `--at-op=@ --ignore-working-copy` to
/// inspect the current state without mutation.
```

Comment (ilyagr) [reply cli/src/commands/operation/log.rs:19]: This text looks good to me, thank you!

Comment (ilyagr) [reply cli/src/commands/operation/log.rs:19]: Although, it still bothers me that if the merge of concurrent operations is failing, there's no way to see both concurrent operations at once, since `--at-op @` will give an error. Hopefully that would be rare?

Comment (yuja) [reply cli/src/commands/operation/log.rs:19]: > Hopefully that would be rare?

I believe it is. Typical use case is `op log` or `op log --ignore-working-copy`, and concurrent ops get merged. `--at-op=@` is needed only when merge fails (i.e. repository is somewhat corrupted.)

**Pull request merged**
